File: js/src/utils/ketbra.ts
import { conj, matrix, multiply } from 'mathjs';
import type { Matrix } from 'mathjs';
import type { VectorLike } from '../types.js';
import { toVector } from './internal.js';

/**
 * The outer product |a><b|.
 *
 * @param a - the "ket" vector.
 * @param b - the "bra" vector (its complex conjugate is used). Defaults to `a`.
 * @returns the rank-1 operator |a><b|.
 */
export function ketbra(a: VectorLike, b: VectorLike = a): Matrix {
  const ket = matrix(toVector(a).map((entry) => [entry])); // column, n x 1
  const bra = conj(matrix([toVector(b)])); // conjugated row, 1 x m
  return multiply(ket, bra);
}

File: python/src/bound_entangled/utils/ketbra.py
import numpy as np


def ketbra(a: np.ndarray, b: np.ndarray | None = None) -> np.ndarray:
    """The outer product |a><b|.

    Args:
        a: the "ket" vector.
        b: the "bra" vector (its complex conjugate is used).

    Returns:
        np.ndarray: the rank-1 operator |a><b|.
    """
    if b is None:
        b = a
    return np.outer(a, b.conj())
